h/services/annotation.py

```diff
@@ -52,6 +61,90 @@ def search_annotations(
+        :param update_timestamp: Whether to update the last-edited timestamp of
+            the annotation.
+        :param reindex_tag: Tag used by the reindexing job to identify the
+            source of the reindexing request.
+        """
+        initial_target_uri = annotation.target_uri
+
+        self._update_annotation_values(annotation, data)
+        if update_timestamp:
+            annotation.updated = datetime.utcnow()
+
+        # Expire the group relationship, so we get the most up-to-date value
+        # instead of the one which was present when we loaded the model
+        # https://docs.sqlalchemy.org/en/13/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7
+        self._db.expire(annotation, ["group"])
+        self._validate_group(annotation)
+
+        if (
+            document := data.get("document", {})
+        ) or annotation.target_uri != initial_target_uri:
+            annotation.document = update_document_metadata(
+                self._db,
+                annotation.target_uri,
+                document.get("document_meta_dicts", {}),
+                document.get("document_uri_dicts", {}),
+                updated=annotation.updated,
+            )
+
+        # The search index service by default does not reindex if the existing ES
+        # entry's timestamp matches the DB timestamp. If we're not changing this
+        # timestamp, we need to force reindexing.
+        self._search_index_service._queue.add_by_id(  # pylint: disable=protected-access
+            annotation.id, tag=reindex_tag, schedule_in=60, force=not update_timestamp
+        )
+
+        return annotation
+
+    @staticmethod
+    def _update_annotation_values(annotation: Annotation, data: dict):
+        for key, value in data.items():
+            # Don't set complex things
+            if key in ("document", "extra"):
+                continue
+
+            setattr(annotation, key, value)
+
+        extra = data.get("extra", {})
+        annotation.extra.update(extra)
+
+    @staticmethod
+    def _validate_group(annotation: Annotation):
+        group = annotation.group
+        if not group:
+            raise ValidationError(
+                "group: " + _(f"Invalid group id {annotation.groupid}")
+            )
+
+        # If no scopes are present, or if the group is configured to allow
+        # annotations outside its scope, there's nothing to do here
+        if not group.scopes or not group.enforce_scope:
+            return
+
+        # The target URI must match at least one
+        # of a group's defined scopes, if the group has any
+        if not url_in_scope(
+            annotation.target_uri, [scope.scope for scope in group.scopes]
+        ):
+            raise ValidationError(
+                "group scope: "
+                + _("Annotations for this target URI are not allowed in this group")
+            )
```

Comment:
This is a different formulation than the original because when done this way it can be re-used for `create_annotation`